kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java

```diff
@@ -341,78 +317,108 @@ private SnapshotImpl createSnapshot(
+   *       checkpoint version.
+   *   <li>Third, process and validate this list of _delta_log files to yield a {@code LogSegment}.
+   * </ol>
    */
-  private Optional<LogSegment> getLogSegmentFrom(
-      Engine engine, Optional<CheckpointMetaData> startingCheckpoint) {
-    return getLogSegmentForVersion(
-        engine, startingCheckpoint.map(x -> x.version), Optional.empty());
+  private LogSegment getLogSegmentForVersion(Engine engine, Optional<Long> versionToLoadOpt) {
+    final String versionToLoadStr = versionToLoadOpt.map(String::valueOf).orElse("latest");
+    logger.info("Loading log segment for version {}", versionToLoadStr);
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////
+    // Step 1: Find the latest checkpoint version. If $versionToLoadOpt is empty, use the version //
+    //         referenced by the _LAST_CHECKPOINT file. If $versionToLoad is present, search for  //
+    //         the previous latest complete checkpoint at or before $versionToLoad.               //
+    ////////////////////////////////////////////////////////////////////////////////////////////////
+
+    final Optional<Long> startCheckpointVersionOpt =
+        versionToLoadOpt
+            .map(
+                versionToLoad -> {
+                  logger.info(
+                      "Finding last complete checkpoint at or before version {}", versionToLoad);
+                  final long startTimeMillis = System.currentTimeMillis();
+                  return Checkpointer.findLastCompleteCheckpointBefore(
+                          engine, logPath, versionToLoad + 1)
+                      .map(checkpointInstance -> checkpointInstance.version)
+                      .map(
+                          checkpointVersion -> {
+                            checkArgument(
+                                checkpointVersion <= versionToLoad,
+                                "Last complete checkpoint version %s was not <= targetVersion %s",
+                                checkpointVersion,
+                                versionToLoad);
+
+                            logger.info(
+                                "{}: Took {}ms to find last complete checkpoint at or before "
+                                    + "targetVersion {}",
+                                tablePath,
+                                System.currentTimeMillis() - startTimeMillis,
+                                versionToLoad);
+
+                            return checkpointVersion;
+                          });
+                })
+            .orElseGet(
+                () -> {
+                  logger.info("Loading last checkpoint from the _last_checkpoint file");
+                  return new Checkpointer(logPath)
+                      .readLastCheckpointFile(engine)
+                      .map(x -> x.version);
+                });
```

Comment:
Q: in this case (where `versionToLoadOpt` is empty), didn't we previously default to using `findLastCompleteCheckpointBefore` if `_last_checkpoint` didn't exist. Wondering if this is a meaningful change.

Reply:
Good question. We did not do that. `findLastCompleteCheckpointBefore` is only useful if you can give it a version to search for _before_. If versionToLoad is empty, then we don't have an end version to compare against.

https://github.com/delta-io/delta/blob/107b2d3c3279093ab6481503c399b828579a031a/kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java#L381